docs/source/literal_types.rst

```diff
@@ -168,6 +161,89 @@ For example, compare and contrast what happens when you try appending these type
     reveal_type(list_of_lits)  # Revealed type is 'List[Literal[19]]'
 
 
+Intelligent indexing
+********************
+
+We can use Literal types to more precisely index into structured heterogeneous
+types such as tuples, NamedTuples, and TypedDicts. This feature is known as
+*intelligent indexing*.
+
+For example, when we index into a tuple using some int, the inferred type is
+normally the union of the tuple item types. However, if we want just the type
+corresponding to some particular index, we can use Literal types like so:
+
+.. code-block:: python
+
+    tup = ("foo", 3.4)
+
+    # Indexing with an int literal gives us the exact type for that index
+    reveal_type(tup[0])  # Revealed type is 'str'
+
+    # But what if we want the index to be a variable? Normally mypy won't
+    # know exactly what the index is and so will return a less precise type:
+    int_index = 1
+    reveal_type(tup[int_index])  # Revealed type is 'Union[str, float]'
+
+    # But if we use either Literal types or a Final int, we can gain back
+    # the precision we originally had:
+    lit_index: Literal[1] = 1
+    fin_index: Final = 1
+    reveal_type(tup[lit_index])  # Revealed type is 'str'
+    reveal_type(tup[fin_index])  # Revealed type is 'str'
```

Comment:
Maybe add an example with TypedDict? Maybe also an example with union of literals, something along these lines:
```python
i: Literal[1, 2]
t: Tuple[str, int, int]
reveal_type(t[i])  # This is "int"
```